File: objective-c/tutorial5/tutorial5/ViewController.m
#import "ViewController.h"
@import RMQClient;

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self receiveLogsTopic:@[@"kern.*", @"*.critical"]];
    sleep(2);
    [self emitLogTopic:@"Hello World!" routingKey:@"kern.info"];
    [self emitLogTopic:@"A critical kernel error" routingKey:@"kern.critical"];
    [self emitLogTopic:@"Critical module error" routingKey:@"somemod.critical"];
    [self emitLogTopic:@"Just some module info. You won't get this." routingKey:@"somemod.info"];
}

- (void)receiveLogsTopic:(NSArray *)routingKeys {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];
    RMQExchange *x    = [ch topic:@"topic_logs"];
    RMQQueue *q       = [ch queue:@"" options:RMQQueueDeclareExclusive];

    for (NSString *routingKey in routingKeys) {
        [q bind:x routingKey:routingKey];
    }

    NSLog(@"Waiting for logs.");

    [q subscribe:^(RMQMessage * _Nonnull message) {
        NSLog(@"%@:%@", message.routingKey, [[NSString alloc] initWithData:message.body encoding:NSUTF8StringEncoding]);
    }];
}

- (void)emitLogTopic:(NSString *)msg routingKey:(NSString *)routingKey {
    RMQConnection *conn = [[RMQConnection alloc] initWithDelegate:[RMQConnectionDelegateLogger new]];
    [conn start];

    id<RMQChannel> ch = [conn createChannel];
    RMQExchange *x    = [ch topic:@"topic_logs"];

    [x publish:[msg dataUsingEncoding:NSUTF8StringEncoding] routingKey:routingKey];
    NSLog(@"Sent '%@'", msg);

    [conn close];
}

@end

File: swift/tutorial5/tutorial5/ViewController.swift
//
//  ViewController.swift
//  tutorial5
//
//  Copyright © 2016 RabbitMQ. All rights reserved.
//
import UIKit
import RMQClient

class ViewController: UIViewController {


    override func viewDidLoad() {
        super.viewDidLoad()
        self.receiveLogsTopic(["kern.*", "*.critical"])
        sleep(2)
        self.emitLogTopic("Hello World!", routingKey: "kern.info")
        self.emitLogTopic("A critical kernel error", routingKey: "kern.critical")
        self.emitLogTopic("Critical module error", routingKey: "somemod.critical")
        self.emitLogTopic("Just some module info. You won't get this.", routingKey: "somemod.info")
    }

    func receiveLogsTopic(_ routingKeys: [String]) {
        let conn = RMQConnection(delegate: RMQConnectionDelegateLogger())
        conn.start()
        let ch = conn.createChannel()
        let x = ch.topic("topic_logs")
        let q = ch.queue("", options: .exclusive)
        for routingKey: String in routingKeys {
            q.bind(x, routingKey: routingKey)
        }
        print("Waiting for logs.")
        q.subscribe({(_ message: RMQMessage) -> Void in
            print("\(message.routingKey!):\(String(data: message.body, encoding: .utf8)!)")
        })
    }

    func emitLogTopic(_ msg: String, routingKey: String) {
        let conn = RMQConnection(delegate: RMQConnectionDelegateLogger())
        conn.start()
        let ch = conn.createChannel()
        let x = ch.topic("topic_logs")
        x.publish(msg.data(using: .utf8), routingKey: routingKey)
        print("Sent '\(msg)'")
        conn.close()
    }
}
